### PR TITLE
Configuration for (converting) Populators

### PR DESCRIPTION
**Pull request by mike4git**
A next step into a generalized configuration for common Converters AND Populators

### CODE REVIEW

File: config/packages/neusta_converter.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+neusta_converter:
+  converter:
+    test.person.converter:
+#      converter: Neusta\ConverterBundle\Converter\GenericConverter
+      target_factory: Neusta\ConverterBundle\Tests\Fixtures\Model\PersonFactory
+      populators:
+        - Neusta\ConverterBundle\Tests\Fixtures\Populator\PersonNamePopulator
+#      properties:
```
Comment (lukadschaak) [inline config/packages/neusta_converter.yaml:8]: Sollen diese auskommentierten Codezeilen drinnen bleiben?

Comment (mike4git) [reply config/packages/neusta_converter.yaml:8]: Ja, ich glaube die dienen als "Vorlage" sozusagen.

Comment (lukadschaak) [reply config/packages/neusta_converter.yaml:8]: Okay, Sounds like a case for docs for me :)
But nevermind, this is okay for me.

**Review by lukadschaak — APPROVED** [review]

File: config/packages/neusta_converter.yaml
Comment (jdreesen) [inline config/packages/neusta_converter.yaml:1]: What is this configuration file for? It is not used anywhere. It seems like a copy from the one of the test app.

File: src/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -68,4 +70,41 @@ private function addConverterSection(ArrayNodeDefinition $rootNode): void
...
+                                ->useAttributeAsKey('target')
+                                ->prototype('scalar')
+                                ->isRequired()
+                            ->end()
+                        ->end()
+                    ->end()
+                    ->validate()
+                        ->ifTrue(fn (array $c) => empty($c['converter']) && empty($c['property']))
+                        ->thenInvalid('A "converter" and "property" must be defined.')
+                    ->end()
```
Comment (jdreesen) [inline src/DependencyInjection/Configuration.php:105]: Do we really need this? Both settings are required anyway.

File: src/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -68,4 +70,41 @@ private function addConverterSection(ArrayNodeDefinition $rootNode): void
             ->end()
         ;
     }
+
+    private function addPopulatorSection(ArrayNodeDefinition $rootNode): void
+    {
+        $rootNode
+            ->children()
+                ->arrayNode('populators')
```
Comment (jdreesen) [inline src/DependencyInjection/Configuration.php:78]: Can we change this to `populator`? The `converter` setting is also singular.

File: src/DependencyInjection/NeustaConverterExtension.php [hunk 4]
```diff
@@ -67,6 +74,63 @@ private function registerConverterConfiguration(string $id, array $config, Conta
...
+     * @param array<string, mixed> $config
+     */
+    private function registerPopulatorConfiguration(string $id, array $config, ContainerBuilder $container): void
+    {
+        $arguments = [];
+        if (empty($config['class']) || ConvertingPopulator::class === $config['class']) {
+            $arguments = $this->buildArgumentsForConvertingPopulator($config);
+        } elseif (ArrayConvertingPopulator::class === $config['class']) {
+            $arguments = $this->buildArgumentsForArrayConvertingPopulator($config);
+        }
```
Comment (jdreesen) [inline src/DependencyInjection/NeustaConverterExtension.php:87]: This is a good case for the new `match` expression:

```suggestion
        $arguments = match ($config['class']) {
            ConvertingPopulator::class => $this->buildArgumentsForConvertingPopulator($config),
            ArrayConvertingPopulator::class => $this->buildArgumentsForArrayConvertingPopulator($config),
            default => [],
        };
```

(also add `->cannotBeEmpty()` to the `class` config node)

File: src/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -68,4 +70,41 @@ private function addConverterSection(ArrayNodeDefinition $rootNode): void
...
+                    ->info('Populator configuration')
+                    ->normalizeKeys(false)
+                    ->useAttributeAsKey('name')
+                    ->arrayPrototype()
+                        ->children()
+                            ->scalarNode('class')
+                                ->info('class of the "Populator" implementation')
+                                ->defaultValue(Populator\ConvertingPopulator::class)
+                            ->end()
```
Comment (jdreesen) [inline src/DependencyInjection/Configuration.php:87]: Should we name this `populator` to be consistent with the `converter` config?

File: src/DependencyInjection/NeustaConverterExtension.php [hunk 6]
```diff
@@ -67,6 +74,63 @@ private function registerConverterConfiguration(string $id, array $config, Conta
...
+    /**
+     * @param array<string, mixed> $config
+     *
+     * @return array<string, string>
+     */
+    private function buildArgumentsForConvertingPopulator(array $config): array
+    {
+        $targetProperty = array_key_first($config['property']);
+        $sourceProperty = $config['property'][$targetProperty];
+        $sourceProperty = $sourceProperty ?? $targetProperty;
```
Comment (jdreesen) [inline src/DependencyInjection/NeustaConverterExtension.php:103]:
```suggestion
        $sourceProperty = $config['property'][$targetProperty] ?? $targetProperty;
```

File: src/DependencyInjection/NeustaConverterExtension.php [hunk 7]
```diff
@@ -67,6 +74,63 @@ private function registerConverterConfiguration(string $id, array $config, Conta
...
+
+        return
+            [
+                '$converter' => new TypedReference($config['converter'], Converter::class),
+                '$sourcePropertyName' => $sourceProperty,
+                '$targetPropertyName' => $targetProperty,
+                '$accessor' => new Reference('property_accessor'),
+            ];
```
Comment (jdreesen) [inline src/DependencyInjection/NeustaConverterExtension.php:111]:
```suggestion
        return [
            '$converter' => new TypedReference($config['converter'], Converter::class),
            '$sourcePropertyName' => $sourceProperty,
            '$targetPropertyName' => $targetProperty,
            '$accessor' => new Reference('property_accessor'),
        ];
```

File: src/DependencyInjection/NeustaConverterExtension.php [hunk 8]
```diff
@@ -67,6 +74,63 @@ private function registerConverterConfiguration(string $id, array $config, Conta
...
+        $innerPropertyArgument['$sourceArrayItemPropertyName'] = $innerProperty;
+        unset($config['property']['itemProperty']);
+
+        $arguments = array_merge(
+            $innerPropertyArgument,
+            $this->buildArgumentsForConvertingPopulator($config),
+        );
+
+        return $arguments;
+    }
```
Comment (jdreesen) [inline src/DependencyInjection/NeustaConverterExtension.php:132]: This is not correct. The parameter names of the two classes are different.

File: tests/DependencyInjection/NeustaConverterExtensionTest.php [hunk 9]
```diff
@@ -123,6 +126,122 @@ public function test_with_mapped_context(): void
...
+                ],
+            ],
+        ]);
+
+        // converter
+        $populator = $container->getDefinition('foobar');
+
+        self::assertSame(ConvertingPopulator::class, $populator->getClass());
+        self::assertTrue($populator->isPublic());
+        self::assertInstanceOf(TypedReference::class, $populator->getArgument('$converter'));
```
Comment (jdreesen) [inline tests/DependencyInjection/NeustaConverterExtensionTest.php:147]: Can we also assert that the value is `GenericConverter::class`?

File: tests/DependencyInjection/NeustaConverterExtensionTest.php [hunk 10]
```diff
@@ -123,6 +126,122 @@ public function test_with_mapped_context(): void
...
+        self::assertSame('targetTest', $populator->getArgument('$targetPropertyName'));
+        self::assertSame('sourceTest', $populator->getArgument('$sourcePropertyName'));
+    }
```
Comment (jdreesen) [inline tests/DependencyInjection/NeustaConverterExtensionTest.php:173]: This doesn't test the `ArrayConvertingPopulator`. It's exactly the same as the one above it.

File: tests/DependencyInjection/NeustaConverterExtensionTest.php [hunk 11]
```diff
@@ -123,6 +126,122 @@ public function test_with_mapped_context(): void
...
+    public function test_with_array_converting_populator_with_inner_property(): void
+    {
+        $container = $this->buildContainer([
+            'populators' => [
+                'foobar' => [
+                    'class' => ArrayConvertingPopulator::class,
+                    'converter' => GenericConverter::class,
+                    'property' => [
+                        'targetTest' => 'sourceTest',
+                        'itemProperty' => 'value',
```
Comment (jdreesen) [inline tests/DependencyInjection/NeustaConverterExtensionTest.php:184]: I don't think this config is good UX. 

The first entry is `target property name` => `source property name` (custom key with custom value) and the second is `config key name` => `config key value` (specific config key with custom value), so it's mixing two different kinds of configuration here.

**Review by jdreesen — COMMENTED** [review]
Could you please describe this new feature in the docs?

File: config/packages/neusta_converter.yaml
Comment (jdreesen) [reply config/packages/neusta_converter.yaml:1]: See #55

File: src/DependencyInjection/Configuration.php (hunk 2 above)
Comment (jdreesen) [reply src/DependencyInjection/Configuration.php:105]: See #55

File: src/DependencyInjection/Configuration.php (hunk 3 above)
Comment (jdreesen) [reply src/DependencyInjection/Configuration.php:78]: See #55

File: src/DependencyInjection/NeustaConverterExtension.php (hunk 4 above)
Comment (jdreesen) [reply src/DependencyInjection/NeustaConverterExtension.php:87]: See #55

File: src/DependencyInjection/Configuration.php (hunk 5 above)
Comment (jdreesen) [reply src/DependencyInjection/Configuration.php:87]: See #55

File: src/DependencyInjection/NeustaConverterExtension.php (hunk 6 above)
Comment (jdreesen) [reply src/DependencyInjection/NeustaConverterExtension.php:103]: See #55

File: src/DependencyInjection/NeustaConverterExtension.php (hunk 7 above)
Comment (jdreesen) [reply src/DependencyInjection/NeustaConverterExtension.php:111]: See #55

File: src/DependencyInjection/NeustaConverterExtension.php (hunk 8 above)
Comment (jdreesen) [reply src/DependencyInjection/NeustaConverterExtension.php:132]: See #55

File: tests/DependencyInjection/NeustaConverterExtensionTest.php (hunk 9 above)
Comment (jdreesen) [reply tests/DependencyInjection/NeustaConverterExtensionTest.php:147]: See #55

File: tests/DependencyInjection/NeustaConverterExtensionTest.php (hunk 10 above)
Comment (jdreesen) [reply tests/DependencyInjection/NeustaConverterExtensionTest.php:173]: See #55

File: tests/DependencyInjection/NeustaConverterExtensionTest.php (hunk 11 above)
Comment (jdreesen) [reply tests/DependencyInjection/NeustaConverterExtensionTest.php:184]: See #56

**Pull request merged**